website/docs/proposals/20250620-auto-scaler-for-inference-workloads.md

```diff
@@ -0,0 +1,468 @@
+				return CurrentReplicas
+			}
+		}
+
+		// check cooldown
+		if cooldownElapsed > DownCoolDownSeconds {
+			return max(CurrentReplicas - ScaleDownStep, MinReplicas)
+		}
+	}
+
+	// 4. otherwise, skip scaling action
+	return CurrentReplicas
+}
+
+func filterMetricsWithinWindow(key, metricsHistory, stabilizationWindowSeconds) windowMetrics {
+	windowStartTime := time.Now().Add(-time.Second * time.Duration(stabilizationWindowSeconds))
+
+	for i, timestampMetric := range metricsHistory[key] {
+		if timestampMetric.timestamp.After(windowStartTime) {
+			windowMetrics = append(windowMetrics, timestampMetric.Value)
+		}
+	}
+
+	return windowMetrics
+}
+
+func updateMetricsHistory(key, currentValue, metricsHistory, upStabilizationWindowSeconds, downStabilizationWindowSeconds) {
+	upWindowStartTime := time.Now().Add(-time.Second * time.Duration(upStabilizationWindowSeconds))
+	downWindowStartTime := time.Now().Add(-time.Second * time.Duration(downStabilizationWindowSeconds))
+
+	foundStaleMetric := false
+	staleMetricIndex := 0
+	for i, timestampMetric := range metricsHistory[key] {
+		if timestampMetric.timestamp.Before(upWindowStartTime) && timestampMetric.timestamp.Before(downWindowStartTime) {
+			foundStaleMetric = true
+			staleMetricIndex = i
+		}
+	}
+
+	// use the stale metric slot for controlling the size of metric slice
+	if foundStaleMetric {
+		metricsHistory[key][staleMetricIndex] = timestampedMetricValue{currentValue, time.Now()}
+	} else {
+		metricsHistory[key] = append(metricsHistory[key], timestampedMetricValue{currentValue, time.Now())
+	}
+}
+```
+
+### Cron Scaler
+
+The Cron Scaler provides a proactive, time-based scaling mechanism that adjusts the number of replicas at scheduled times. Unlike the Realtime Scaler, which reacts to metric changes, the Cron Scaler allows you to align resource allocation with predictable traffic patterns, such as scaling up before business hours and scaling down during nights or weekends. This approach ensures that capacity is available before demand spikes, improving responsiveness and user experience.
+
+Each cron policy defines a `schedule` in standard cron format and a `targetReplicas` count. The autoscaler will continuously monitor the clock and, when the current time matches a policy's schedule, it will adjust the workload's replica count to the specified target.
+
+- Take the following policies as an example:
+
+```yaml
+policies:
+  # The "scale-up" policy increases the number of replicas to 5 at 8:00 AM every day
+  # to handle peak traffic during business hours.
+  - name: "scale-up"
+    schedule: "0 8 * * *" # Cron schedule for 8:00 AM daily
+    targetReplicas: 5
+  # The "scale-down" policy reduces the number of replicas to 1 at 1 minute past midnight (00:01)
+  # every day to conserve resources during off-peak hours.
+  - name: "scale-down"
+    schedule: "1 0 * * *" # Cron schedule for 00:01 AM daily
+    targetReplicas: 1
+```
```

Comment:
Is a scale-down policy enforced when a scale-up policy is created, what happens when the user only creates scale-up policy but not scale-down policy?

Reply:
There are no enforced policy settings in current proposal. If no scale-down policy is configured, the replicas will be equal to MaxReplicas after some scaling actions.